src/tooltip.tsx

```diff
@@ -135,10 +136,77 @@ const Tooltip: React.FC<Props> = ({
         height: 0,
     });
 
-    const getPosition = (position: Position = defaultPositionDesktop) =>
-        isTabletOrSmaller && (position === 'left' || position === 'right') ? defaultPositionMobile : position;
+    const [containerPosition, setContainerPosition] = React.useState({});
+
+    const getPosition = React.useCallback(
+        (position: Position | undefined) =>
+            isTabletOrSmaller ? position || defaultPositionMobile : position || defaultPositionDesktop,
+        [isTabletOrSmaller]
+    );
+
+    const [tooltipClientRect, setTooltipClientRect] = React.useState<DOMRect | undefined>(undefined);
+
+    const tooltipBoundingClientRect = React.useCallback(() => {
+        if (!tooltipRef.current) return undefined;
+
+        return tooltipRef.current.getBoundingClientRect();
+    }, []);
+
+    const hasRightSpace = React.useCallback(() => {
+        if (!tooltipClientRect) return undefined;
+
+        return targetBoundingClientRect.current.right + tooltipClientRect.width > window.innerWidth;
+    }, [tooltipClientRect]);
+
+    const hasLeftSpace = React.useCallback(() => {
+        if (!tooltipClientRect) return undefined;
+
+        return targetBoundingClientRect.current.left < tooltipClientRect.width;
+    }, [tooltipClientRect]);
+
+    const validatePosition = React.useCallback(
```

Comment:
you can remove all the `useCallbacks`. They are not needed unless you want to pass the resulting function to a child component to avoid re-renders